### PR TITLE
Added functional tests for ModelMetadataType attribute

### PR DESCRIPTION
**Pull request by kirthik**

### CODE REVIEW

**Comment by kirthik** [comment]
@dougbu , @rynowak - Could you please take a look?

File: src/Microsoft.AspNet.Mvc.ModelBinding/Validation/DataAnnotationsModelValidatorProvider.cs [hunk 1]
```diff
@@ -61,7 +61,11 @@ private static bool AddImplicitRequiredAttributeForValueTypes
                 {
                     factory = _defaultAttributeFactory;
                 }
-                results.Add(factory(attribute));
+                var modelValidator = factory(attribute);
+                if (!results.Exists(a => a.GetType() == modelValidator.GetType()))
+                {
+                    results.Add(factory(attribute));
+                }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Validation/DataAnnotationsModelValidatorProvider.cs]: - why execute the delegate (`factory`) twice?
- is this a fix for a problem you found while testing?  should be done in a separate commit w/ clear comments in any case.

File: test/WebSites/ModelBindingWebSite/Models/Software.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Mvc;
+using System;
+using System.ComponentModel.DataAnnotations;
```
Comment (dougbu) [inline test/WebSites/ModelBindingWebSite/Models/Software.cs]: System namespaces first

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTests.cs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+using Microsoft.AspNet.TestHost;
+using Newtonsoft.Json;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.FunctionalTests
+{
+    public class ModelMetadataAttributeTests
+    {
+        private readonly IServiceProvider _services = TestHelper.CreateServices(nameof(ModelBindingWebSite));
+        private readonly Action<IApplicationBuilder> _app = new ModelBindingWebSite.Startup().Configure;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTests.cs]: suggest creating a new `ValidationWebSite` since validation relates to model binding but is not the same sub-system.  the tests for `TryValidateModel` would also fit in the new `ValidationWebSite`.

File: test/WebSites/ModelBindingWebSite/Models/ProductValidatorAttibute.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    public class ProductValidatorAttribute : ValidationAttribute
+    {
+        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+        {
+            var objectType = value.GetType();
+
+            var neededProperties =
+              objectType.GetProperties()
+              .Where(propertyInfo => propertyInfo.Name == "Name" || propertyInfo.Name == "Country")
+              .ToArray();
```
Comment (dougbu) [inline test/WebSites/ModelBindingWebSite/Models/ProductValidatorAttibute.cs]: similar to the admin validator in the `TryValidateModel` PR, conditionally cast to `Product`.  fail if that returns `null`.  then will be able to use the properties directly.

File: test/WebSites/ModelBindingWebSite/Models/ProductValidatorAttibute.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace ModelBindingWebSite
```
Comment (dougbu) [inline test/WebSites/ModelBindingWebSite/Models/ProductValidatorAttibute.cs]: nit: move outside Models directory

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTests.cs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.TestHost;
+using Newtonsoft.Json;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.FunctionalTests
+{
+    public class ModelMetadataAttributeTests
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTests.cs]: we're inconsistent in following this rule but the [Engineering Guidelines](https://github.com/aspnet/Home/wiki/Engineering-guidelines) say
> Test class names end with Test

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTests.cs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            Assert.Equal("CompanyName cannot be null", json["product.CompanyName"]);
+            Assert.Equal("The field Price must be between 20 and 100.", json["product.Price"]);
+            Assert.Equal("The Category field is required.", json["product.Category"]);
+            Assert.Equal("The ContactUs field is required.", json["product.Contact"]);
+            Assert.Equal("The Field2 field is required.", json["product.ProductDetails.Field2"]);
+            Assert.Equal("The Field3 field is required.", json["product.ProductDetails.Field3"]);
+        }
+
+        [Fact]
+        public async Task InValidComplexTypeOnBaseClass_Product()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTests.cs]: nit: "Invalid" (lowercase 'v')

File: test/WebSites/ModelBindingWebSite/Models/CompanyNameAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.DataAnnotations;
+
+namespace ModelBindingWebSite
```
Comment (dougbu) [inline test/WebSites/ModelBindingWebSite/Models/CompanyNameAttribute.cs]: nit: `ModelBindingWebSite.Models`

Comment (dougbu) [inline test/WebSites/ModelBindingWebSite/Models/CompanyNameAttribute.cs]: oops. this class doesn't belong in this directory

File: test/WebSites/ModelBindingWebSite/Models/Product.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.DataAnnotations;
+
+namespace ModelBindingWebSite
```
Comment (dougbu) [inline test/WebSites/ModelBindingWebSite/Models/Product.cs]: `ModelBindingWebSite.Models` for this and other classes that belong (and remain) in this directory

File: test/WebSites/ModelBindingWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+                {
+                    var value = error.Exception != null ? error.Exception.Message :
+                                                          error.ErrorMessage;
+                    result.Add(item.Key, value);
+                }
+            }
+
+            return result;
+        }
+
```
Comment (dougbu) [inline test/WebSites/ModelBindingWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: nit: extra blank line

File: test/WebSites/ModelBindingWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+using System.Linq;
+using Microsoft.AspNet.Mvc;
+using ModelBindingWebSite.ViewModels;
+
+namespace ModelBindingWebSite.Controllers
+{
+    public class ModelMetadataTypeAttributeController : Controller
+    {
+        [HttpPost]
+        public object ValidateProductViewModelInclMetadata([FromBody]ProductViewModel product)
```
Comment (dougbu) [inline test/WebSites/ModelBindingWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: nit: space between attribute and parameter

File: test/WebSites/ModelBindingWebSite/Models/CompanyNameAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public class CompanyNameAttribute : ValidationAttribute
+    {
+        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+        {
+            object instance = validationContext.ObjectInstance;
```
Comment (dougbu) [inline test/WebSites/ModelBindingWebSite/Models/CompanyNameAttribute.cs]: `var`

**Comment by dougbu** [comment]
:watch:
there's a lot here and I'm hoping the final versions of this and the `TryValidateModel` PR can share a site and the model / attribute classes in that site.

File: test/WebSites/ValidationWebSite/ProductValidatorAttibute.cs [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+namespace ValidationWebSite
+{
+    public class ProductValidatorAttribute : ValidationAttribute
+    {
+        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+        {
+
+            if (value is ProductViewModel)
+            {
+                var product = (ProductViewModel)value;
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/ProductValidatorAttibute.cs]: nit: demonstrate correct use of conditional and unconditional casts:
``` c#
  var product = value as ProductViewModel;
  if (product != null)
  {
    ...
  }
  else
  {
    var software = (SoftwareViewModel)value;
    ..
  }
```

could be extra-careful and use `as` for SoftwareViewModel as well. then return a `ValidationResult` explaining the validator didn't get an instance of the expected type.  right now, that case will throw an `InvalidCastException`.

File: test/WebSites/ValidationWebSite/ProductValidatorAttibute.cs [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+                if (!product.Country.Equals("USA") && string.IsNullOrEmpty(product.Name))
+                {
+                    return new ValidationResult("Country and Name fields don't have the right values");
+                }
+
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/ProductValidatorAttibute.cs]: nit: remove extra blank line

File: test/WebSites/ValidationWebSite/ProductValidatorAttibute.cs [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+{
+    public class ProductValidatorAttribute : ValidationAttribute
+    {
+        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+        {
+
+            if (value is ProductViewModel)
+            {
+                var product = (ProductViewModel)value;
+                if (!product.Country.Equals("USA") && string.IsNullOrEmpty(product.Name))
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/ProductValidatorAttibute.cs]: not clear why products need to be made in the USA only if they aren't named.  should this condition use `||`?

File: test/WebSites/ValidationWebSite/ProductValidatorAttibute.cs [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+                    return new ValidationResult("Country and Name fields don't have the right values");
+                }
+
+            }
+            else
+            {
+                var software = (SoftwareViewModel)value;
+                if (!software.Country.Equals("USA") && string.IsNullOrEmpty(software.Name))
+                {
+                    return new ValidationResult("Country and Name fields don't have the right values");
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/ProductValidatorAttibute.cs]: this is a functional test but others will use the code as demonstrations of best practises.  make the message actionable and end it with a period; something like "Product must be made in the USA if it is not named."

File: test/WebSites/ValidationWebSite/Models/Product.cs [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+namespace ValidationWebSite.Models
+{
+    [ProductValidator]
+    public class Product
+    {
+        public string Name { get; set; }
+
+        [StringLength(20)]
+        [RegularExpression("^[0-9]*$")]
+        [Display(Name = "ContactUs")]
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Models/Product.cs]: since this is for display, "Contact Us" (add a space)

File: test/WebSites/ValidationWebSite/CompanyNameAttribute.cs [hunk 18]
```diff
@@ -3,19 +3,16 @@
...
 {
-    public class AdminValidator : ValidationAttribute
+    public class CompanyNameAttribute : ValidationAttribute
     {
         protected override ValidationResult IsValid(object value, ValidationContext validationContext)
         {
-            var admin = (Administrator)value;
-            
-            if (admin.AdminAccessCode != 1)
+            if (value == null)
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/CompanyNameAttribute.cs]: nit: should this ensure the `value` is a `string` that's not `null` or empty?

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+using System.Linq;
+using Microsoft.AspNet.Mvc;
+using ValidationWebSite.ViewModels;
+
+namespace ValidationWebSite.Controllers
+{
+    public class ModelMetadataTypeValidationController : Controller
+    {
+        [HttpPost]
+        public object ValidateProductViewModelInclMetadata([FromBody] ProductViewModel product)
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: don't abbreviate: "Incl" -> "With" or "Including"

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+                }
+                if (errorMessage != string.Empty)
+                {
+                    result.Add(item.Key, errorMessage);
+                }
+            }
+
+            return result;
+        }
+
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: nit: remove extra blank line

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+        [HttpGet]
+        public object TryValidateModelSoftwareViewModelNoPrefix()
+        {
+            var softwareVm = new SoftwareViewModel
+            {
+                Category = "Technology",
+                CompanyName = "Microsoft",
+                Contact = "4258393231",
+                Country = "UK",
+                DatePurchased = new DateTime(10, 10, 10),
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: nit: move the purchase to the last few years, 2010 perhaps

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+            }
+            TryValidateModel(product);
+
+            return CreateValidationDictionary();
+        }
+
+        [HttpGet]
+        public object TryValidateModelSoftwareViewModelNoPrefix()
+        {
+            var softwareVm = new SoftwareViewModel
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: don't abbreviate: "Vm" -> "ViewModel"

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+            {
+                Category = "Technology",
+                CompanyName = "Microsoft",
+                Contact = "4258393231",
+                Country = "USA",
+                DatePurchased = new DateTime(10, 10, 10),
+                Name = "MVC",
+                Price = 110,
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: nit: MVC is free as in beer

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+            {
+                var errorMessage = string.Empty;
+                foreach (var error in item.Value.Errors)
+                {
+                    if (error != null)
+                    {
+                        errorMessage = errorMessage + error.ErrorMessage;
+                    }
+                }
+                if (errorMessage != string.Empty)
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: `!string.IsNullOrEmpty()` (with very, very few exceptions and this isn't one of them)

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        [HttpPost]
+        public object TryValidateModelProductViewModelWithErrorInParameter(int id, [FromBody] ProductViewModel product)
+        {
+            if (id == 0)
+            {
+                ModelState["id"].Errors.Clear();
+            }
+            TryValidateModel(product);
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: are you demonstrating that `TryValidateModel` dos not add an `Error` related to missing `id` value?  if yes, it's not clear from the action name.  besides that, there's no requirement the leaves the `id` value out.  suggest at least adding some comments about what this action demonstrates.

could also use `[Required, MaxLength(3), StringLength(10) { MinLength = 5 }] string theImpossibleString, ...` to ensure client _can't_ send a valid value for the parameter.

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+                Version = "2"
+            };
+
+            TryValidateModel(softwareVm);
+
+            return CreateValidationDictionary();
+        }
+
+        [HttpGet]
+        public object TryValidateModelValidModelNoPrefix()
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: both this action and `TryValidateModelSoftwareViewModelNoPrefix` call `TryValidateModel()` without a prefix.  either change this one to use a prefix or add another action (and corresponding test).

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTest.cs [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+
+namespace Microsoft.AspNet.Mvc.FunctionalTests
+{
+    public class ModelMetadataAttributeTest
+    {
+        private readonly IServiceProvider _services = TestHelper.CreateServices(nameof(ValidationWebSite));
+        private readonly Action<IApplicationBuilder> _app = new ValidationWebSite.Startup().Configure;
+
+        [Fact]
+        public async Task ValidBaseClass_Product()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTest.cs]: rename all test methods to include expectations; follow the `method_conditions_expectations` or `method_expectations_conditions` pattern (preferably, pick one :smiley_cat:)

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTest.cs [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+            // Arrange
+            var server = TestServer.Create(_services, _app);
+            var client = server.CreateClient();
+            var input = "{ \"Name\": \"MVC\", \"Contact\":\"4258959019\", \"Category\":\"Technology\"," +
+                "\"CompanyName\":\"Microsoft\", \"Country\":\"USA\",\"Price\": 21, \"ProductDetails\": {\"Detail1\": \"d1\"," +
+                " \"Detail2\": \"d2\", \"Detail3\": \"d3\"}}";
+            var content = new StringContent(input, Encoding.UTF8, "application/json");
+
+            var url =
+                "http://localhost/ModelMetadataTypeValidation/ValidateProductViewModelInclMetadata";
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTest.cs]: nit: don't wrap this line if it is already short enough

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs [hunk 29]
```diff
@@ -14,60 +16,75 @@ namespace Microsoft.AspNet.Mvc.FunctionalTests
...
     public class TryValidateModelTest
     {
-        private readonly IServiceProvider _services = TestHelper.CreateServices(nameof(FormatterWebSite));
-        private readonly Action<IApplicationBuilder> _app = new FormatterWebSite.Startup().Configure;
+        private readonly IServiceProvider _services = TestHelper.CreateServices(nameof(ValidationWebSite));
+        private readonly Action<IApplicationBuilder> _app = new ValidationWebSite.Startup().Configure;
 
         [Fact]
-        public async Task TryValidateModel_SimpleModelInvalidProperties()
+        public async Task InvalidProperties_ModelStateContainsErrorBeforecallingTryValidateModel()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: nits:
- "Calling" (capital 'C')
- likely needs another underscore somewhere in the name

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: also, what are you expecting here?

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs [hunk 30]
```diff
@@ -14,60 +16,75 @@ namespace Microsoft.AspNet.Mvc.FunctionalTests
...
         {
             // Arrange
             var server = TestServer.Create(_services, _app);
             var client = server.CreateClient();
+            var input = "{ \"Price\": 2, \"Contact\": \"acvrdzersaererererfdsfdsfdsfsdf\", "+
+                "\"ProductDetails\": {\"Detail1\": \"d1\", \"Detail2\": \"d2\", \"Detail3\": \"d3\"}}";
+            var content = new StringContent(input, Encoding.UTF8, "application/json");
+            var url =
+                "http://localhost/ModelMetadataTypeValidation/TryValidateModelProductViewModelWithErrorInParameter/0";
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: will this lead to a model state error for `id`?  if not, how does the parameter on the action matter?

**Comment by dougbu** [comment]
:watch:

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs (hunk 25 above)
Comment (kirthik) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: Idea here was so clear a ModelState error and make sure TryValidateModel does not add it back somehow.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs (hunk 30 above)
Comment (kirthik) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: yes, it will lead to model state error on id

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs (hunk 26 above)
Comment (kirthik) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: added prefix for `TryValidateModelSoftwareViewModelNoPrefix `.

File: test/WebSites/ValidationWebSite/CompanyNameAttribute.cs (hunk 18 above)
Comment (kirthik) [inline test/WebSites/ValidationWebSite/CompanyNameAttribute.cs]: This logs a validation error when value is null. I cannot ensure it is a string when value is null right? For cases where value is not null, I am anyways not doing anything and returning null.

Comment (dougbu) [inline test/WebSites/ValidationWebSite/CompanyNameAttribute.cs]: to test if `value` is a non-`null` `string`, use two statements e.g. `var valueString = value as string; if (string.IsNullOrEmpty(valueString) ...`

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs (hunk 25 above)
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: this action and the associated test remain unclear.  what error would the `ModelState["id"].Errors.Clear()` remove?  why is the removal conditional -- no test sends anything but `0`?

I suggested the `[Required, MaxLength(3), StringLength(10) { MinLength = 5 }] string theImpossibleString` approach to make sending a completely valid request impossible.  the action should then start with `ModelState["theImpossibleString"].Errors.Clear();` and otherwise ignore the parameter.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs [hunk 31]
```diff
@@ -14,60 +16,75 @@ namespace Microsoft.AspNet.Mvc.FunctionalTests
...
     public class TryValidateModelTest
     {
-        private readonly IServiceProvider _services = TestHelper.CreateServices(nameof(FormatterWebSite));
-        private readonly Action<IApplicationBuilder> _app = new FormatterWebSite.Startup().Configure;
+        private readonly IServiceProvider _services = TestHelper.CreateServices(nameof(ValidationWebSite));
+        private readonly Action<IApplicationBuilder> _app = new ValidationWebSite.Startup().Configure;
 
         [Fact]
-        public async Task TryValidateModel_SimpleModelInvalidProperties()
+        public async Task TryValidateModel_InvalidProperties_ModelStateContainsErrorBeforeCallToTryValidateModel()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: this name describes the test arrangement, not expectations

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTest.cs [hunk 32]
```diff
@@ -0,0 +1,188 @@
...
+        {
+            // Arrange
+            var server = TestServer.Create(_services, _app);
+            var client = server.CreateClient();
+
+            var url = "http://localhost/ModelMetadataTypeValidation/ValidateSoftwareViewModelIncludingMetadata";
+            var request = new HttpRequestMessage(HttpMethod.Post, url);
+
+            var version = "2.2";
+            request.Headers.TryAddWithoutValidation("version", version);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelMetadataAttributeTest.cs]: not clear how this header affects the test.  if it's not necessary, change the test to use `PostAsync()` and otherwise be more similar to other tests.  if it's required, add code comments explaining the need.

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 33]
```diff
@@ -0,0 +1,100 @@
...
+        [HttpPost]
+        public object ValidateSoftwareViewModelIncludingMetadata([FromBody] SoftwareViewModel software)
+        {
+            return CreateValidationDictionary();
+        }
+
+        [HttpPost]
+        public object TryValidateModelProductViewModelWithErrorInParameter(int id, [FromBody] ProductViewModel product)
+        {
+            //Clear ModelState entry if id is 0. TryValidateModel should not add entries that are cleared.
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: always include a space after `//`

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 34]
```diff
@@ -0,0 +1,100 @@
...
+            {
+                ModelState["id"].Errors.Clear();
+            }
+            TryValidateModel(product);
+
+            return CreateValidationDictionary();
+        }
+
+        [HttpGet]
+        public object TryValidateModelSoftwareViewModelNoPrefix()
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: action name is no longer correct; should be `TryValidateModelSoftwareViewModelWithPrefix`

**Comment by dougbu** [comment]
:watch: close but has a couple of unclear aspects

**Comment by kirthik** [comment]
@dougbu - addressed your comments. Could you please take a look?

File: test/WebSites/ValidationWebSite/CompanyNameAttribute.cs [hunk 35]
```diff
@@ -3,17 +3,16 @@
...
         protected override ValidationResult IsValid(object value, ValidationContext validationContext)
         {
-            var admin = (Administrator)value;
-            
-            if (admin.AdminAccessCode != 1)
+            var valueString = value as string;
+            if (string.IsNullOrEmpty(valueString))
             {
-                return new ValidationResult ("AdminAccessCode property does not have the right value");
+                return new ValidationResult("CompanyName cannot be null.");
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/CompanyNameAttribute.cs]: teensy nit: "... null or empty"

File: test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs [hunk 36]
```diff
@@ -0,0 +1,101 @@
...
+        {
+            return CreateValidationDictionary();
+        }
+
+        [HttpPost]
+        public object TryValidateModelAfterClearingValidationErrorInParameter(
+            [Required, MaxLength(3), StringLength(10, MinimumLength = 3)] string theImpossibleString,
+            [FromBody] ProductViewModel product)
+        {
+            // Clear ModelState entry. TryValidateModel should not add entries that are cleared.
```
Comment (dougbu) [inline test/WebSites/ValidationWebSite/Controllers/ModelMetadataTypeAttributeController.cs]: nit: "that are cleared" -> "except those found within the passed model"

**Comment by dougbu** [comment]
:shipit:
